src/bm_sim/actions.cpp

```diff
@@ -289,12 +312,13 @@ ActionFnEntry::execute(Packet *pkt) const {
 
   auto &primitives = action_fn->primitives;
   size_t param_offset = 0;
-  // primitives is a vector of pointers
   BMLOG_TRACE_SI_PKT(*pkt, action_fn->get_source_info(),
                      "Executing action {}", action_fn->get_name());
-  for (auto primitive : primitives) {
-    primitive->execute(&state, &(action_fn->params[param_offset]));
-    param_offset += primitive->get_num_params();
+  for (size_t idx = 0; idx < primitives.size();) {
+    const auto &primitive = primitives[idx];
+    param_offset = primitive.get_param_offset();
+    primitive.execute(&state, &(action_fn->params[param_offset]));
+    idx = primitive.get_jump_offset(idx);
```

Comment:
it does not have to be, we could imagine supporting backward jumps (for loops)
what would make sense is to check the jump indices in the bmv2 json parser but that would require some special handling for the jump primitives, which I am trying to avoid